### ISSUE TITLE
Bug loss calculation: 0. in pred follows in NaNs in bbox_iou(...)

### ISSUE THREAD

**Issue by UnglvKitDe** [ultralytics/yolov5#8539]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Training, Validation

### Bug

The problem occurs when the prediction has 0.0 values. These then follow through the method bbox_iou in NANs in the loss calculation.
https://github.com/ultralytics/yolov5/blob/526e650553819dbff67897b9c752c4072e989823/train.py#L352-L353
https://github.com/ultralytics/yolov5/blob/526e650553819dbff67897b9c752c4072e989823/utils/loss.py#L138-L142
In the example below, the NaNs result here:
https://github.com/ultralytics/yolov5/blob/526e650553819dbff67897b9c752c4072e989823/utils/metrics.py#L247
This is one reason for such a training: 
![results](https://user-images.githubusercontent.com/100289696/178155368-4b0bda05-03df-4658-9be1-0416bb4c656b.png)

If you agree with me, I'll make a PR :) 

### Environment

YOLOv5 torch 1.11 (cuda 11.3) and 1.12 (cuda 11.6)

### Minimal Reproducible Example

```
pxy = torch.tensor([[0.2729, 0.6113],
        [0.6572, 0.6045],
        [0.2646, 0.6406],
        [0.6221, 0.6357],
        [0.3359, 0.2222],
        [0.7158, 0.4521]], device='cuda:0', dtype=torch.float16,
       requires_grad=True)
pwh = torch.tensor([[7.3277e-08, 5.9422e-07],
        [0.0000e+00, 0.0000e+00],
        [7.5654e-08, 9.5031e-07],
        [0.0000e+00, 0.0000e+00],
        [1.5166e-06, 1.4725e-05],
        [3.2162e-11, 0.0000e+00]], device='cuda:0', requires_grad=True)
pbox = torch.cat((pxy, pwh), 1)
tbox = torch.tensor([[ 0.6705,  0.7127,  4.0370,  2.7977],
        [ 0.6705,  0.7127,  4.0370,  2.7977],
        [-0.3295,  0.7127,  4.0370,  2.7977],
        [-0.3295,  0.7127,  4.0370,  2.7977],
        [ 0.6705, -0.2873,  4.0370,  2.7977],
        [ 0.6705, -0.2873,  4.0370,  2.7977]], device='cuda:0')
iou = bbox_iou(pbox, tbox, CIoU=True).squeeze()  # iou(prediction, target)
print(iou) # tensor([-0.0711,     nan, -0.0914,     nan, -0.0851, -0.0420])
```

### Additional

One solution would be to add `torch.finfo(torch.float32).tiny`

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@UnglvKitDe very interesting! These should be protected divides on metrics.py L247. I will take a look.

**Comment by UnglvKitDe** [comment]
As written [here](https://discuss.pytorch.org/t/nan-loss-with-torch-cuda-amp-and-crossentropyloss/108554/18) we could cast pred (float16) to float32 (target is also float32). Otherwise you might get problems because of AMP.  Then we can also add torch.finfo(torch.float32).tiny to all or only to the 0s.

**Comment by glenn-jocher** [comment]
@UnglvKitDe we have an `eps=1e-7` value here that's added to h1 and h2 for xyxy format boxes, it looks like a recent update missed adding eps to the h1 and h2 variables created by xywh boxes:
https://github.com/ultralytics/yolov5/blob/526e650553819dbff67897b9c752c4072e989823/utils/metrics.py#L216-L230

**Comment by UnglvKitDe** [comment]
@glenn-jocher Ah that's a good point, I didn't see that thanks!

**Comment by glenn-jocher** [comment]
@UnglvKitDe I opened up PR #8546 to fix this issue. Can you please review? Thanks!

**Comment by UnglvKitDe** [comment]
@glenn-jocher Done. Thank you! :)

**Comment by glenn-jocher** [comment]
@UnglvKitDe you're welcome! Thank you for your contribution and for helping improve YOLOv5! If you have any more issues or questions, feel free to ask.
